src/ucp/wireup/wireup_cm.c

```diff
@@ -1079,10 +1127,23 @@ static ssize_t ucp_cm_server_priv_pack_cb(void *arg,
  */
 static unsigned ucp_cm_server_conn_notify_progress(void *arg)
 {
-    ucp_ep_h ucp_ep = arg;
+    ucp_ep_h ucp_ep               = arg;
+    ucp_wireup_ep_t *cm_wireup_ep = ucp_ep_get_cm_wireup_ep(ucp_ep);
+    ucs_status_t status;
 
     UCS_ASYNC_BLOCK(&ucp_ep->worker->async);
-    ucp_wireup_remote_connected(ucp_ep);
+    /* TCP don't have CONNECT_TO_EP support and has internal connection
+     * matching that could lead to unexpected behavior when connections
+     * are accepted in the reverse order.
+     * TODO: remove it, when CONNECT_TO_EP support is added to TCP */
+    if (!strcmp(ucp_context_cm_name(ucp_ep->worker->context,
+                                    cm_wireup_ep->cm_idx), "tcp")) {
+        ucp_wireup_remote_connected(ucp_ep);
+    } else {
+        ucp_ep->flags |= UCP_EP_FLAG_LISTENER;
```

Comment:
why need to add UCP_EP_FLAG_LISTENER flag, where are we setting ucp_ep_ext_gen(ep)->listener ?